### PR TITLE
ENH: Truncated weibull (minimum) distribution

### PR DESCRIPTION
**Pull request by ali-cetin-4ss**
This pull-request add a new distribution function: left truncated Weibull distribution. This distribution functions is useful when describing "general" exponential tail behavior.

This distribution function exhibits similar difficulties  as Weibull 3-p when fitting to data. A naive MLE is not well-behaved (4 parameters, where uniqueness is not always possible). Therefore, a modified MLE functions is applied for when the truncation parameter needs to be estimated.

### CODE REVIEW

File: scipy/stats/_continuous_distns.py
Comment (ilayn) [inline scipy/stats/_continuous_distns.py]: Since Python 2 is dropped you can just write `return (a**c - sc.log1p(-q))**(1/c)`

File: scipy/stats/_continuous_distns.py
Comment (ilayn) [inline scipy/stats/_continuous_distns.py]: Can you explain why you want to have context manager here? If it is for scoping you don't need it.

File: scipy/stats/_continuous_distns.py
Comment (ilayn) [inline scipy/stats/_continuous_distns.py]: I think you can just write a bigger try block instead of jumping through the `super` and context managers here. It is basically two-three lines each for the branches.

File: scipy/stats/_continuous_distns.py
Comment (ilayn) [inline scipy/stats/_continuous_distns.py]: `**kwargs` as a convention

File: scipy/stats/_continuous_distns.py
Comment (ilayn) [inline scipy/stats/_continuous_distns.py]: Default for `get` is already `None` but you are not using the actual values anyways so you don't need to get the values. Instead you can convert these to explicit bools 

```Python
return any([x in kwds for x in ['f0', 'fa', 'fix_a']])
```
is a definite True or False then below you can just check 

```Python
if f0 or (floc  and fscale):
    ...
```

which I think is much more readable.

**Review by ilayn — COMMENTED** [review]
Thanks @ace-e4s for taking the time for this PR. I've left some generic comments but a stats resident would check the functional parts better than I can.

**Comment by ace-e4s** [comment]
@ilayn , thanks for the feedback. I shall review them as soon as possible.
(Taking time to contribute to scipy is the least I can do to :) )

**Comment by ilayn** [comment]
No rush take your time :)

**Comment by chrisb83** [comment]
Could you add a reference for this distribution? Thanks

**Comment by ace-e4s** [comment]
> Could you add a reference for this distribution? Thanks

Sure. This is just vanilla left-truncated 3-p Weibull distribution. Some good text-books on the subject may be:
1. The Weibull Distribution: A Handbook, Horst Rinne, CRC press
2. Statistical methods for Reliability data, W.Q. Meeker and L. A. Escobar, Wiley, 1998, pp. 266- 

However, the challenge with a 3-p left truncated Weibull distribution is that ML functions is not well-behaved when all 3p + truncation parameter are to be estimated from data. Uniqueness is not guaranteed, e.g. c = 1, then a and loc are interchangeable. The usual challenges with 3p Weibull come on top of that. (Usually, the truncation value is know a priori, but the scipy framework is a littlebit cumbersome here. see Issue https://github.com/scipy/scipy/issues/9439).

The proposed implementation attempts to work around the issues by estimating the truncation parameter in a non-parametric fashion. Then find the remaining parameters with MLE. (I saw this somewhere, but trying to find the ref.) However, if one of the parameters are locked, then the usual MLE is used.

Furthermore, the current start values provided by scipy.stats framework don't work very well for this distribution. (open for suggestions here.) Therefore, it is crucial to provide good start value for at least the scale parameter if all parameters are kept free.

File: .gitignore
Comment (ace-e4s) [reply .gitignore]: Oopss! unintentionally added. I will revert.

File: scipy/stats/__init__.py
Comment (ace-e4s) [reply scipy/stats/__init__.py]: I was just trying to follow the convention "trunc" + "distribution name". Since the new distribution is the truncation of "weibull_min"... well, "truncweibull_min".

Anyway, I think the call is yours (scipy core dev team). I have no particular preference.

File: scipy/stats/_continuous_distns.py
Comment (ace-e4s) [reply scipy/stats/_continuous_distns.py]: agreed. I will make the change.

File: scipy/stats/_continuous_distns.py
Comment (ace-e4s) [reply scipy/stats/_continuous_distns.py]: Agree in general, but as PEP8 suggests: consistency in module first. This module uses kwds all over the place, so did I.

Comment (ilayn) [reply scipy/stats/_continuous_distns.py]: Ah then my mistake, I was probably to fast reviewing. Nevermind my comment

File: scipy/stats/_continuous_distns.py
Comment (ace-e4s) [reply scipy/stats/_continuous_distns.py]: That is true for "f0". However, just trying to be consistent with the rest of the code base here, e.g. see lognorm_gen.fit().

For floc (and fscale) it is necessary to explicitly check for None, since 0 is a valid input.

File: scipy/stats/_continuous_distns.py
Comment (ace-e4s) [reply scipy/stats/_continuous_distns.py]: Good question. I think I was experimenting with something more complex, then simplified it and did not realize that it could be eliminated all together.

File: scipy/stats/_continuous_distns.py
Comment (ace-e4s) [reply scipy/stats/_continuous_distns.py]: I have now removed the context manager and moved it all into a `try/finally` statement. However, I think you still need to call `super`...?

**Comment by chrisb83** [comment]
Thanks for the references. So the truncated Weibul distribution is just a Weibull distribution conditioned on the inteval [a, b], where you consider the special case b=inf (left truncation), no?

Would it be easy to implement the general case with right and left truncation as special cases?

**Comment by ace-e4s** [comment]
> Thanks for the references. So the truncated Weibul distribution is just a Weibull distribution conditioned on the inteval [a, b], where you consider the special case b=inf (left truncation), no?
> Would it be easy to implement the general case with right and left truncation as special cases?

Yes, you are right. The special case b=inf is often useful when it is desired to describe some "general" exponential tail. 

It should be straight forward to define cdf, pdf etc for double truncated Weibull. However, parameter estimation is likely to be a nightmare, if practically feasible at all if all parameters are set free.

**Comment by chrisb83** [comment]
> However, parameter estimation is likely to be a nightmare, if practically feasible at all if all parameters are set free.

OK, I understand. One could probably implement the double truncated Weibull and a fit method that works well for the special case that b = inf, whereas the general case relies on the generic fit method (which might not work great).

**Comment by ace-e4s** [comment]
> 
> 
> > However, parameter estimation is likely to be a nightmare, if practically feasible at all if all parameters are set free.
> 
> OK, I understand. One could probably implement the double truncated Weibull and a fit method that works well for the special case that b = inf, whereas the general case relies on the generic fit method (which might not work great).

That's a viable solution. Let me sleep on for couple of days and do some literature research on doubly truncated weibull... (After all, when you want to use truncated distribution functions, it's because you know the the truncation point. In those cases, the standard routine may work well.)

**Comment by ace-e4s** [comment]
I've been looking at double truncated Weibull distribution as agreed. But I have run into a "numerical accuracy" issue.. Therefore, had to introduce a function: `_log1mexp(x)`. Let me demonstrate this with the CDF only:

```
def _log1mexp(x):
    """
    Compute log(1 - exp(-x))

    Choose optimal numerical formulation for accuracy.

    References
    ----------
    1. Accurately Computing log(1 − exp(−|a|)) Assessed by the Rmpfr package,
       Martin Machler, ETH Zurih, 2012
    """

    def func_0(x):
        return sc.log1p(-np.exp(-x))

    def func_1(x):
        return np.log(-sc.expm1(-x))

    return _lazyselect([x > np.log(2), x <= np.log(2)], [func_0, func_1], (x, ))

class truncweibull_min_gen(rv_continuous): 
    
   def _argcheck(self, a, b, c):
        self.a = a
        self.b = b
        return (a >= 0.) & (b > 0.) & (c > 0.)

    def _cdf(self, x, a, b, c):
        return sc.expm1(-x**c + a**c)/sc.expm1(-b**c + a**c)

    def _logcdf(self, x, a, b, c):
        return _log1mexp(x**c - a**c) - _log1mexp(b**c - a**c)
```
The difference can be observed by comparing the native logcdf function with the naive one (log(cdf)).

My question is; Is there a  `_log1mexp(x)` equivalent function in numpy/scipy? If not, is it ok to introduce it as a private function or method? Any thoughts?

**Comment by chrisb83** [comment]
> My question is; Is there a _log1mexp(x) equivalent function in numpy/scipy? If not, is it ok to introduce it as a private function or method? Any thoughts?

I am not aware of an existing implementation. You can add it as a private function to the distribution, that is not a problem. one can think about a better place later.

**Comment by ace-e4s** [comment]
I have finally(!) managed to redo the PR to expand it to doubly truncated weibull distribution. I have also refrained from defining a "special" case fit method. Therefore, the default ML will perform very poorly for the general case, but will work for many special cases. (Essentially, you need to fix 2-3 of 5 parameters). This is a known challenge with this distribution as MLE is not well-defined for the general case.

**Comment by ilayn** [comment]
This looks good to me and I don't have other comments as far as I can understand but a `stats` residents should visit this PR.

**Comment by ace-e4s** [comment]
My only concern is related to numerical accuracy for some cases, since we cannot use special.expm1. But, I guess that's just the way it is.

**Comment by mdhaber** [comment]
Looks interesting. I'm going to close and open to re-trigger CI.

File: scipy/stats/__init__.py
Comment (mdhaber) [inline scipy/stats/__init__.py:120]: There are a few other things that need to be done on the [checklist](http://scipy.github.io/devdocs/dev/contributor/adding_new.html#adding-a-new-statistics-distribution).

One doctest error would be fixed by updating the number of continuous distributions in the tutorial document.
The other would be fixed by adding parameters to `_distr_params.py`

File: scipy/stats/_continuous_distns.py
Comment (ev-br) [inline scipy/stats/_continuous_distns.py:2103]: This statement could be clarified. What is the support of `truncweibull_min(c, a=1, b=3, loc=4)`: is it `-3 < x < -1`  --- so that in the formulas above `u_l = -3 = a - loc` and `u_e = -1 = b - loc`?

File: scipy/stats/_continuous_distns.py
Comment (ev-br) [inline scipy/stats/_continuous_distns.py:2146]: `log1p(-q)`, here and in the `_ppf` below

File: scipy/stats/_continuous_distns.py
Comment (ev-br) [inline scipy/stats/_continuous_distns.py:2159]: If you want the reference to be user visible, then add it to the class docstring.

Comment (ace-e4s) [reply scipy/stats/_continuous_distns.py:2159]: Not that important. More useful for reviewers since the expressions for the moments are not that trivial.

File: scipy/stats/_continuous_distns.py
Comment (ace-e4s) [reply scipy/stats/_continuous_distns.py:2146]: I'm not able to see how you can do that (mathematically)? Could you be so provide the full expression as a suggestion? (Note that a summation is going on inside the log)

File: scipy/stats/_continuous_distns.py
Comment (ace-e4s) [reply scipy/stats/_continuous_distns.py:2103]: Well, from the formulas above you see that `u_l = a*scale + loc` and `u_r = b*scale + loc`. Therefore, the support of `truncweibull_min(c, a=1, b=3, loc=4)` is `5 < x <= 7`. Nonetheless, I'll try to add one or two clarifying sences to the dosctring. Feel free to suggest something.

File: scipy/stats/_continuous_distns.py
Comment (ev-br) [reply scipy/stats/_continuous_distns.py:2146]: Indeed. Should have looked closer to the expression to see the closing bracket.

**Review by ilayn — APPROVED** [review]

**Comment by ilayn** [comment]
Looks good to me. So if any stats resident can have a final look, I guess it would be ready.

File: scipy/stats/__init__.py
Comment (ace-e4s) [reply scipy/stats/__init__.py:120]: The above issues should be fixed now.

**Comment by ace-e4s** [comment]
The PR request should be fine now... all checks pass (but appears to be some issue with the mac checks. These passed before).

However, I have done some changes in the tests:
* I had to add this distribution to skip `fails_cmplx` in `test_continuous_basic.py`. The issue seems to be the way clipping is done in `common_tests.py` line 259. This ought to be a issue for all truncated distributions since `a` and `b` are set during function calls.
* I had to add this distribution to `failing_fits` in `test_fit.py. This is expected since the MLE for this distribution is not well defined. Works well only when 2-3 parameters are fixed.

File: scipy/stats/__init__.py
Comment (mdhaber) [reply scipy/stats/__init__.py]: I agree with that reasoning, @ace-e4s.

File: scipy/stats/_continuous_distns.py
Comment (mdhaber) [reply scipy/stats/_continuous_distns.py]: Looks like this is no longer really part of the code; resolving.

File: scipy/stats/_continuous_distns.py
Comment (mdhaber) [reply scipy/stats/_continuous_distns.py]: Yeah looks like the `fit` override is gone.

File: scipy/stats/__init__.py
Comment (mdhaber) [reply scipy/stats/__init__.py:145]: Yes, looks like it.

File: scipy/stats/_continuous_distns.py
Comment (mdhaber) [reply scipy/stats/_continuous_distns.py]: GitHub shows this as outdated, so I'll resolve, but I might add a comment like it to the latest version. This is going to be tough for users, I think. I see why you had to do it this way, though.

File: scipy/stats/_continuous_distns.py
Comment (mdhaber) [reply scipy/stats/_continuous_distns.py:2241]: I think we should add it to the class docstring. I can take care of it.

**Comment by mdhaber** [comment]
> However, I have done some changes in the tests:

Those aren't too concerning to me. Does reference [1] talk about how to fit a one-sided truncated version? If so, maybe someone can submit a follow-up implementing the improved fit strategy for this special case.

File: scipy/stats/_continuous_distns.py
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:2200]: Just want to check - since `_logpdf` is defined, is this explicit formulation of `_pdf` preferable to `np.exp(_logpdf(...))`? Same question for `_logcdf`.

File: scipy/stats/_distr_params.py
Comment (mdhaber) [inline scipy/stats/_distr_params.py:109]: This is not the best choice for the right end of the support because the distribution is essentially only left-truncated.

image

How about:

```suggestion
    ['truncweibull_min', (2.5, 0.25, 1.25)],
```

File: scipy/stats/tests/test_distributions.py
Comment (mdhaber) [inline scipy/stats/tests/test_distributions.py:5000]: How about:

```suggestion
        assert_allclose(ls, -100)

    def test_compare_weibull_min2(self):
        # Verify that the truncweibull_min distribution PDF and CDF results 
        # are the same as those calculated from truncating weibull_min
        c, a, b = 2.5, 0.25, 1.25
        x = np.linspace(a, b, 100)
        
        pdf1 = stats.truncweibull_min.pdf(x, c, a, b)
        cdf1 = stats.truncweibull_min.cdf(x, c, a, b)
        
        norm = stats.weibull_min.cdf(b, c) - stats.weibull_min.cdf(a, c)
        pdf2 = stats.weibull_min.pdf(x, c) / norm
        cdf2 = (stats.weibull_min.cdf(x, c) - stats.weibull_min.cdf(a, c))/norm
        
        np.testing.assert_allclose(pdf1, pdf2)
        np.testing.assert_allclose(cdf1, cdf2)


class TestRdist:
```

File: scipy/stats/_continuous_distns.py
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:2200]:
```suggestion
        return (c * pow(x, c-1) * np.exp(-pow(x, c))) / denom
```
Here and throughout, assuming `num` is "numerator" and `denom` is "denominator". NBD; just for future thought.

**Review by mdhaber — COMMENTED** [review]
~~Test failures are mostly due to MM fit check, which can be disabled.~~ _Done_

~~There is some light linting to be done:~~ _Done_
```
scipy/stats/tests/test_distributions.py:4919:80: E501 line too long (90 > 79 characters)
scipy/stats/tests/test_distributions.py:4923:80: E501 line too long (90 > 79 characters)
scipy/stats/tests/test_distributions.py:4927:80: E501 line too long (90 > 79 characters)
scipy/stats/tests/test_distributions.py:4931:80: E501 line too long (90 > 79 characters)
```

I've checked that the PDF and CDF are correct (in the general case, not just a=0 and b=np.inf) against `weibull_min`, and `_argcheck` and `_get_support` look good. There are enough consistency checks in `scipy.stats` that the rest must be correct.

~~I should still take a look at the tutorial and the `_munp` method line by line.~~ _Checked PDF and CDF, and there are consistency checks on `_munp`, so I'll trust that the author got the LaTeX right in the rest of the tutorial._

A few comments are included inline. I think `**` would be preferable to `pow` throughout, `denom` would be preferable to `denum`, but I suppose that doesn't matter much. Not going to change it or hold this up further for those things.

I'm wondering whether it would make sense to just add truncation bounds as parameters for all distributions and have a generic implementation, since the PDF and CDF of truncated distributions are quite simple transformations of the PDF and CDF of non-truncated distributions. Then we could just override methods (e.g. moment) where we can get substantial accuracy or speed improvements. Other than `_munp`, how much do we gain from the custom implementations here?

File: scipy/stats/_continuous_distns.py
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:2179]: If this is in math formatting, it would be better if `loc` and `scale` were assigned symbols, but I will let it go because it will complicate the text representation, and that upsets others.

File: scipy/stats/_distr_params.py
Comment (mdhaber) [reply scipy/stats/_distr_params.py:109]: Fixed in next commit.

File: scipy/stats/tests/test_distributions.py
Comment (mdhaber) [reply scipy/stats/tests/test_distributions.py:5000]: Added in next commit.

File: scipy/stats/_continuous_distns.py
Comment (mdhaber) [reply scipy/stats/_continuous_distns.py]: Done.

**Comment by mdhaber** [comment]
Ideas for improvement:
* Custom fit method for special cases in which only one side is truncated. Sounds like that's doable. Added this as a comment in gh-11782.
* Seems like we could add a generic implementation of truncated distributions and most methods (pdf/cdf/sf and logs and inverse functions) could be done quite efficiently as long as `cdf` and `ppf` are efficient for the non-truncated distribution. Custom implementations might improve on this, but it seems preferable to giving the truncated distributions different name and writing everything, such as tests for consistency with the original distribution, from scratch. What do you think about that @chrisb83 @ev-br?

**Comment by ev-br** [comment]
> we could add a generic implementation of truncated distributions and most methods 

One issue here is normalization of pdf, which will need to be computed numericall in general, it seems.

**Comment by mdhaber** [comment]
> One issue here is normalization of pdf.

In general, isn't it just divided by `cdf(b) - cdf(a)` (where `cdf` is that of the non-truncated distribution)?

**Comment by ev-br** [comment]
> In general, isn't it just divided by cdf(b) - cdf(a) 

Yeah. Note to self: coffee needs to be taken in-order, before anything else.

**Comment by mdhaber** [comment]
Thanks @ace-e4s and all reviewers!

**Comment by ace-e4s** [comment]
> > One issue here is normalization of pdf.
> 
> @ev-br In general, isn't it just divided by `cdf(b) - cdf(a)` (where `cdf` is that of the non-truncated distribution)? So that's why I wrote that if `cdf` is efficient, than `pdf` will be.

I think that's a great idea. My only concern would be potential numerical issues, which sometimes can be avoided by analytically simplifying the final expression.

**Comment by ace-e4s** [comment]
Thanks for reviewing and approving this PR. I will look into improved fitting strategies and post it in the PR/issue mentioned above.

**Comment by mdhaber** [comment]
> My only concern would be potential numerical issues,

Certainly, and that is why they should be overridden when they can be. The same applies to all the generic implementations (e.g. sf = 1 - cdf stuffers from subtractive cancellation). Those, too, can still be useful while they are waiting to be overridden. Thanks!

**Comment by ace-e4s** [comment]
@mdhaber, one last tiny thing; if/when you work with a generic implementation for truncated distributions, could you please look into how the truncation values are defined. Currently, these are normalized with loc and scale. This creates a catch 22 when you want to set truncation values before fitting to data. In practice, the user cannot use the 'fit' method and is forced to write a wrapper function to minimize the neg likelihood.

**Comment by mdhaber** [comment]
Yes, definitely. It's a _really_ big problem for fitting them, and I also think it's unfortunate in general that we have to ask the users to transform the truncation `a` and `b` based on `loc` and `scale` to use the other methods. We will definitely take these things into consideration when rewriting the distribution superclasses (someday).

**Pull request merged**
